Fix Abilities API tests for WordPress 6.9+

budzanowski opened the pull request:
Fixes Abilities API integration tests for WordPress 6.9+ with backward compatibility.

Fixes WOOAI-112  

Changes:
- Detect WP version by checking for `WP_Ability_Categories_Registry` class
- Use correct action names (`wp_abilities_api_*` for 6.9+, `abilities_api_*` for earlier)
- Set `$wp_actions['init'] = 1` to satisfy WP 6.9's `did_action('init')` check

All 9 tests pass on WP 6.9-beta1.

budzanowski committed:
Fix Abilities API integration tests for WordPress 6.9

WordPress 6.9 bundles the Abilities API in core with breaking changes:
- Action names changed from abilities_api_* to wp_abilities_api_*
- Registry class WP_Abilities_Category_Registry renamed to WP_Ability_Categories_Registry
- get_instance() now requires did_action('init') to return truthy or it returns null

Changes:
- Update all action names to use wp_ prefix (wp_abilities_api_init, wp_abilities_api_categories_init)
- Fake did_action('init') by setting $wp_actions['init'] = 1 in setUp
- Restore init counter after parent::tear_down() to maintain state across tests
- Fix category registry class name reference
- Update .wp-env.json to use WordPress 6.9-beta1 for testing

All 9 tests now pass with 37 assertions.

budzanowski committed:
Add backward compatibility for Abilities API tests

Detect WordPress version by checking for WP_Ability_Categories_Registry class (WP 6.9+ uses plural "Categories", vendor package uses "Category").

Based on detection:
- Use wp_abilities_api_* action names for WP 6.9+
- Use abilities_api_* action names for pre-6.9
- Use correct category registry class name for each version

Keep $wp_actions['init'] = 1 for both versions (doesn't hurt pre-6.9).

Tests now work with both WordPress 6.9+ and earlier versions.

budzanowski committed:
Use proper save/restore pattern for wp_actions init counter

budzanowski committed:
Fix REST API compatibility for WordPress 6.9+

WordPress 6.9+ handles Abilities API REST endpoints differently than the vendor package:
- WP 6.9+ uses core controller classes (WP_REST_Abilities_V1_List_Controller, etc.)
- Earlier versions use vendor package's WP_REST_Abilities_Init class
- Update tests to check for appropriate class based on WordPress version
- Only register vendor REST routes for pre-6.9 versions

budzanowski committed:
restore wp env config

budzanowski committed:
Use version-appropriate REST API routes in tests

WordPress 6.9+ uses different REST API namespace for Abilities API:
- WP 6.9+: /wp-abilities/v1/abilities
- Pre-6.9: /wp/v2/abilities

Update all three REST API tests to use the correct routes based on WordPress version.

timur987 commented on plugins/woocommerce/tests/php/src/Internal/AbilitiesApi/AbilitiesApiIntegrationTest.php:
```diff
@@ -17,14 +17,57 @@ class AbilitiesApiIntegrationTest extends \WC_REST_Unit_Test_Case {
+
+		// Detect WordPress 6.9+ by checking for the core class name (plural "Categories").
+		// WP 6.9+ uses different action names and class names than the vendor package.
+		$is_wp_69_plus = class_exists( 'WP_Ability_Categories_Registry' );
```
I'm thinking whether we should rely on a class name or rely on the specific WP version like
```php
$is_wp_69_plus = version_compare( get_bloginfo( 'version' ), '6.9', '>=' );
```

The name of Categories Registry might be changed, so current implementation might be a bit fragile, but it would definitely work. So, please take it with a grain of salt.

budzanowski replied:
I had some problems with that in the test suite. Will try again.

peterfabian replied:
Or maybe we rename `$is_wp_69_plus` to `$are_abilities_in_wp_core` because that's what we are really interested in?

timur987 commented on plugins/woocommerce/tests/php/src/Internal/AbilitiesApi/AbilitiesApiIntegrationTest.php:
```diff
@@ -321,16 +380,25 @@ function () use ( $ability_id ) {
-		$this->assertTrue( class_exists( 'WP_REST_Abilities_Init' ), 'Bootstrap should load WP_REST_Abilities_Init class' );
+		// Ensure REST API classes are loaded.
+		// WordPress 6.9+ uses core controllers with different namespace, earlier versions use vendor package.
+		$is_wp_69_plus = class_exists( 'WP_Ability_Categories_Registry' );
```
Since we use "is WP 6.9+" check more than once, would it be better to extract it to a separate function?

timur987 reviewed (approved):
Thanks @budzanowski, the code LGTM and since it's for tests, I think is would be safe to go ahead.

Even though everything looks great, I left a couple of minor comments to discuss the way for decision making related to getting a WordPress version. They are non-blockers, just want to know what do you think.

gziolo commented:
I'm working on syncing all updates applied to WordPress core into the `wordpress/abilities-api` Composer package in https://github.com/WordPress/abilities-api/pull/126. Once it gets published and you bump the version, then you probably will have to update the code again.

budzanowski committed:
Extract WordPress 6.9+ version check into helper method

Refactor version detection logic into a reusable is_wp_69_plus() helper method
to improve code maintainability and reduce duplication.

budzanowski committed:
Rename $is_wp_69_plus to $are_abilities_in_wp_core for clarity

This variable name better reflects what we're actually checking for - whether
the Abilities API is bundled in WordPress core, rather than just checking
for a specific version number. This makes the code more semantic and
self-documenting.
